staticaddr/loopin/manager.go

```diff
@@ -0,0 +1,345 @@
+
+	// ValidateLoopInContract validates the contract parameters against our
+	// request.
+	ValidateLoopInContract ValidateLoopInContract
+}
+
+// Manager manages the address state machines.
+type Manager struct {
+	cfg *Config
+
+	runCtx context.Context
+
+	sync.Mutex
+
+	// initChan signals the daemon that the address manager has completed
+	// its initialization.
+	initChan chan struct{}
+
+	// currentHeight stores the currently best known block height.
+	currentHeight uint32
+}
+
+// NewManager creates a new deposit withdrawal manager.
+func NewManager(cfg *Config) *Manager {
+	return &Manager{
+		cfg:      cfg,
+		initChan: make(chan struct{}),
+	}
+}
+
+// Run runs the static address loop-in manager.
+func (m *Manager) Run(ctx context.Context, currentHeight uint32) error {
+	m.runCtx = ctx
+
+	m.Lock()
+	m.currentHeight = currentHeight
+	m.Unlock()
+
+	registerBlockNtfn := m.cfg.ChainNotifier.RegisterBlockEpochNtfn
+	newBlockChan, newBlockErrChan, err := registerBlockNtfn(m.runCtx)
+	if err != nil {
+		return err
+	}
+
+	// Upon start of the loop-in manager we reinstate all previous loop-ins
+	// that are not yet completed.
+	err = m.recover()
+	if err != nil {
+		return err
+	}
+
+	// Communicate to the caller that the address manager has completed its
+	// initialization.
+	close(m.initChan)
+
+	for {
+		select {
+		case height := <-newBlockChan:
+			m.Lock()
+			m.currentHeight = uint32(height)
+			m.Unlock()
+
+		case err = <-newBlockErrChan:
+			return err
+
+		case <-m.runCtx.Done():
+			return m.runCtx.Err()
+		}
+	}
+}
+
+// recover stars a loop-in state machine for each non-final loop-in to pick up
+// work where it was left off before the restart.
+func (m *Manager) recover() error {
+	log.Infof("Recovering static address loop-ins...")
+
+	// Recover loop-ins.
+	loopIns, err := m.cfg.Store.AllLoopIns(m.runCtx)
+	if err != nil {
+		return err
+	}
+
+	for _, loopIn := range loopIns {
+		// If the current deposit is final it wasn't active when we
+		// shut down the client last. So we don't need to start a fsm
+		// for it.
+		if loopIn.IsInFinalState() {
+			continue
+		}
+
+		log.Debugf("Recovering loopIn %x", loopIn.SwapHash[:])
+
+		// Retrieve all deposits regardless of deposit state.
+		// The second return value indicates true iff all returned
+		// deposits are contained in the in-mem active deposits map. On
+		// recovery that is never the case so we can ignore it here.
+		loopIn.Deposits, _ = m.cfg.DepositManager.AllStringOutpointsActiveDeposits( //nolint:lll
+			loopIn.DepositOutpoints, fsm.EmptyState,
+		)
+
+		loopIn.AddressParams, err = m.cfg.AddressManager.GetStaticAddressParameters( //nolint:lll
+			m.runCtx,
+		)
+		if err != nil {
+			return err
+		}
+
+		loopIn.Address, err = m.cfg.AddressManager.GetStaticAddress(
+			m.runCtx,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Create a state machine for a given loop-in.
+		recovery := true
+		fsm, err := NewFSM(m.runCtx, loopIn, m.cfg, recovery)
```

Comment:
We should also store the constructed FSM in a member container, e.g. a map to avoid accidental GC.